packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,591 @@
+        if (t.numEligibleProposals < $.params.numProposalsEnoughForReward) {
+            require(
+                t.lastProposal.creationTimestamp > $.lastProposalRewardsUpdate + $.params.minimumRewardPeriod,
+                'not enough time passed'
+            );
+        }
+        $.lastProposalRewardsUpdate = t.lastProposal.creationTimestamp;
+
+        // Calculate the reward per proposal and per vote
+        t.rewardPerProposal = t.proposalRewardForPeriod / t.numEligibleProposals;
+        t.rewardPerVote = t.votingRewardForPeriod / t.numEligibleVotes;
+
+        //// Second loop over the proposals:
+        //// 1. Skip proposals that were deleted for non eligibility.
+        //// 2. Reward proposal's clientId.
+        //// 3. Reward the clientIds that faciliated voting.
+        //// 4. Make sure all voting clientIds were included.
+
+        InMemoryMapping.Mapping memory m = InMemoryMapping.createMapping({ maxClientId: nextTokenId() - 1 });
+
+        for (uint256 i; i < proposals.length; ++i) {
+            // skip non eligible deleted proposals
+            if (proposals[i].endBlock == 0) continue;
+
+            uint32 clientId = proposals[i].clientId;
+            if (clientId != 0) {
+                m.inc(clientId, t.rewardPerProposal);
+            }
+
+            uint256 votesInProposal;
+            NounsDAOTypes.ClientVoteData[] memory voteData = proposals[i].voteData;
+            uint256 votes;
+            for (uint256 j; j < votingClientIds.length; ++j) {
+                clientId = votingClientIds[j];
+                votes = voteData[j].votes;
+                require(votes > 0, 'all clientId must have votes');
+                if (clientId != 0) {
+                    m.inc(clientId, votes * t.rewardPerVote);
+                }
+                votesInProposal += votes;
+            }
+            require(
+                votesInProposal == proposals[i].forVotes + proposals[i].againstVotes + proposals[i].abstainVotes,
+                'not all votes accounted'
+            );
+        }
+
+        for (uint256 i = 1; i < m.nextAvailableIndex; ++i) {
+            uint256 reward = m.values[i].balance;
+            uint32 clientId = m.values[i].clientId;
+            $._clientBalances[clientId] += reward;
+
+            emit ClientRewarded(clientId, reward);
+        }
+
+        _refundGas(startGas);
+    }
+
+    /**
+     * @notice Withdraws the balance of a client
+     * @dev The caller must be the owner of the NFT with id `clientId`
+     * @dev The maximum value of `amount` is one less than in `_clientBalances[clientId]`.
+     * This in order to leave 1 wei in storage and avoid expensive gas writes in future balance increases.
+     * @param clientId Which client balance to withdraw
+     * @param amount amount of withdraw
+     * @param to the address to withdraw to
+     */
+    function withdrawClientBalance(uint32 clientId, uint256 amount, address to) public whenNotPaused {
```

Comment:
nit: I think it's nicer for `to` to come before `amount` to better mirror the ERC20 transfer interface

Reply:
a bit lazy to do the refactor right now, but noted!

Reply:
@davidbrai I pushed this refactor :)